pkg/controllers/etcd.go

```diff
@@ -46,15 +46,39 @@ func (s *EtcdService) Name() string           { return "etcd" }
 func (s *EtcdService) Dependencies() []string { return []string{} }
 
 func (s *EtcdService) Run(ctx context.Context, ready chan<- struct{}, stopped chan<- struct{}) error {
+	defer close(stopped)
+
+	// Check to see if we should run as a systemd run or directly as a binary.
+	runningAsSvc := os.Getenv("INVOCATION_ID") != ""
+
+	// Get the path to the etcd binary based on the MicroShift binary location.
 	microshiftExecPath, err := os.Executable()
 	if err != nil {
 		return fmt.Errorf("%v failed to get exec path: %v", s.Name(), err)
 	}
 	etcdPath := filepath.Join(filepath.Dir(microshiftExecPath), "microshift-etcd")
+	// Not running the etcd binary directly, the proper etcd arguments are handled
+	//		in etcd/cmd/microshift-etcd/run.go.
 	args := []string{"run"}
 
+	// If we're launching MicroShift as a service, we need to do the same
+	//		with etcd, so wrap it in a transient systemd-unit that's tied
+	//		to the MicroShift service lifetime.
+	var exe string
+	if runningAsSvc {
+		args = append([]string{
+			"--uid=root",
+			"--scope",
+			"--property", "BindsTo=microshift.service",
+			"--unit", "microshift-etcd",
+			etcdPath,
+		}, args...)
+		exe = "systemd-run"
```

Comment:
Is this `systemd-run` execution a blocking one?
Or does it simply starts the unit and exits? - in this case, cmd.Process used in `defer func()` after `cmd.Start()` can cause some misleading logs to appear.

If that's non-blocking, then maybe for transient unit we want `cmd.Run()` instead of `cmd.Start()`?
We'd add `TimeoutStartSec` to unit and `github.com/coreos/go-systemd/daemon.daemon.SdNotify(false, daemon.SdNotifyReady)` (like we have [here](https://github.com/openshift/microshift/blob/main/pkg/cmd/run.go#L142)) to the `microshift-etcd`'s `run.go`.
In such case `cmd.Run()` would run to completion which would be "starting the unit and informing whether it was started successfully".

I know `microshift-etcd` will be tied to `microshift`, but maybe we could use that as a fail safe, rather than default path.
WDYT about shutting down `microshift-etcd` unit gracefully (i.e. issuing stop and waiting for exit) upon microshift's shutdown? I'm not sure what SIG will systemd send to `microshift-etcd`.
Because with etcd logs in different unit we could miss things when looking at `journalctl -u microshift`.

Reply:
systemd-run is non-blocking; it creates/starts the unit and exits. 
> In such case cmd.Run() would run to completion which would be "starting the unit and informing whether it was started successfully".
Yeah this makes sense, I'll make this change.

I'm fine adding a bit more to not rely on systemd to shut us down when shutting down ushift.

Reply:
I'm sorry @dusk125 
I just checked systemd-run docs and it states:
> If a command is run as transient scope unit, (...).
> Execution in this case is synchronous, and will return only when the command finishes. This mode is enabled via the --scope switch (see below).

Coming back to my questions:
> Is this systemd-run execution a blocking one?
> (...)
> If that's non-blocking, then maybe for transient unit we want cmd.Run() instead of cmd.Start()?

So it looks like we don't have the luxury of "reusing systemd unit's timeout" which I meant there.
We just need to treat `systemd-run` just like we'd treat `microshift-etcd`, right?
This means that there's no use of "Run()" because it'll block indefinitely.
Also all what I wrote about `daemon.SdNotify()` is unnecessary (because there's no such mechanism for transient units)

Reply:
So it seems that we can revert changes related to that and just keep wrapping microshift-etcd with systemd... Sorry for that, I should've check that earlier...

Reply:
Will do, thanks!